src/common/innerPlugin.ts

```diff
@@ -132,21 +141,25 @@ export function makeOtherMethods<T>(
   return out
 }
 
-function upgradeMetaTokens(metaTokens: EdgeMetaToken[]): EdgeTokenMap {
-  const out: EdgeTokenMap = {}
-  for (const metaToken of metaTokens) {
-    const { contractAddress } = metaToken
-    if (contractAddress == null) continue
-    out[contractToTokenId(contractAddress)] = {
-      currencyCode: metaToken.currencyCode,
-      denominations: metaToken.denominations,
-      displayName: metaToken.currencyName,
-      networkLocation: { contractAddress: metaToken.contractAddress }
-    }
+/**
+ * Downgrades EdgeToken objects to the legacy EdgeMetaToken format.
+ */
+function makeMetaTokens(tokens: EdgeTokenMap): EdgeMetaToken[] {
+  const out: EdgeMetaToken[] = []
+  for (const tokenId of Object.keys(tokens)) {
+    const { currencyCode, displayName, denominations, networkLocation } =
+      tokens[tokenId]
+
+    const cleanLocation = asMaybeContractLocation(networkLocation)
+    if (cleanLocation == null) continue
```

Comment:
if an plugin doesn't use contractAddress in EdgeMetaToken (xrp doesn't) then you won't get any metatokens. Is this okay? If not, you should probably do the inverse and check if a plugin has metaTokens defined, if so then use them, if not derive them.

Reply:
I have made this up to the plugins to decide. Tron and all the EVM's are calling `metaTokens: makeMetaTokens(builtinTokens)` in their currencyInfo's, but the Ripple plugin can do whatever it needs to make this work.

Reply:
Ok. got it.